lldb/source/Utility/RealpathPrefixes.cpp

```diff
@@ -0,0 +1,66 @@
+//===-- RealpathPrefixes.cpp ----------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "lldb/Utility/RealpathPrefixes.h"
+
+#include "lldb/Target/Statistics.h"
+#include "lldb/Target/Target.h"
+#include "lldb/Utility/FileSpec.h"
+#include "lldb/Utility/FileSpecList.h"
+#include "lldb/Utility/LLDBLog.h"
+#include "lldb/Utility/Log.h"
+
+using namespace lldb_private;
+
+RealpathPrefixes::RealpathPrefixes(const FileSpecList &file_spec_list)
+    : m_fs(llvm::vfs::getRealFileSystem()), m_target(nullptr) {
+  m_prefixes.reserve(file_spec_list.GetSize());
+  for (const FileSpec &file_spec : file_spec_list) {
+    m_prefixes.emplace_back(file_spec.GetPath());
+  }
+}
+
+void RealpathPrefixes::SetFileSystem(
+    llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs) {
+  m_fs = fs;
+}
+
+std::optional<FileSpec>
+RealpathPrefixes::ResolveSymlinks(const FileSpec &file_spec) const {
+  if (m_prefixes.empty())
+    return std::nullopt;
+
+  auto is_prefix = [](llvm::StringRef a, llvm::StringRef b,
+                      bool case_sensitive) -> bool {
+    return case_sensitive ? a.consume_front(b) : a.consume_front_insensitive(b);
```

Comment:
Why don't you care that this time around that you aren't doing this only with full directory names?

Reply:
Are you asking why don't we call `is_prefix()` with the directory, and instead we are calling it with the path (directory + basename) while we know that the prefix should only be for the directory?

Reply:
I was asking more about the difference between this `is_prefix` and the `is_suffix` that you defined when first figuring out whether one of the real path prefixes are applied to the incoming path.

Reply:
Ah got it.

The `is_suffix` and the body of `IsCompatible()` is existing code. I refactored it from originally an inner loop into this function. It seems git wasn't able to figure out which lines are new and which is old.

In general I feel prefixes on partial names generally make sense - ppl sometimes name their directories or files with the same prefix. E.g. we have `lldb/Utility/LLDBAssert.h` and `lldb/Utility/LLDBLog.h`.   But not a big deal, if you think full directory/file names are better/safer, then changing into it is easy.

Reply:
I can't remember offhand what the source path mapping does - that's the most obvious analog.  If that does substitution into partial names when adjusting paths, it seems reasonable this do so as well.  But if it doesn't, then this probably shouldn't either.

Reply:
That's reasonable.

IIUC, the source path logic:
* Say the source-map is: /foo => /bar
* The user set breakpoint for /barqux/main.cpp:1
* It causes `PathMappingList::ReverseRemapPath()`
* STRING prefix match is successful (/bar is prefix of /barqux/main.cpp)
* qux/main.cpp is the remaining list of components
* Each COMPONENT is added onto the other side of the mapping, the /foo, as separate path components
* The result is /foo/qux/main.cpp

Overall I'd say this logic cares about components of paths. (Side note: It's a bit weird to me that the [prefix match](https://github.com/llvm/llvm-project/blob/main/lldb/source/Target/PathMappingList.cpp#L209) is done as a pure string match.)

Let me change `is_prefix` into component-sensitive, too. Probably by checking if the remaining path is empty or starts with a "/", similar to what the `is_suffix` does.